src/topology_sampler.hpp

```diff
@@ -0,0 +1,139 @@
+// Copyright 2019-2021 bito project contributors.
+// bito is free software under the GPLv3; see LICENSE file for details.
+
+// A class that samples one tree topology per call to the Sample() function.
+// The parameters are as follows:
+//   node - a node to start sampling for
+//   dag - the SubsplitDAG that owns "node"
+//   normalized_sbn_parameters - edge probabilities for leafward sampling
+//   inverted_probabilities - edge probabilities for rootward sampling
+
+#pragma once
+
+#include "subsplit_dag.hpp"
+#include "node.hpp"
+#include "mersenne_twister.hpp"
+
+class TopologySampler {
+ public:
+  // Sample a single tree from the DAG according to the provided edge
+  // probabilities.
+  Node::NodePtr Sample(SubsplitDAGNode node, SubsplitDAG& dag,
+                       EigenConstVectorXdRef normalized_sbn_parameters,
+                       EigenConstVectorXdRef inverted_probabilities);
+
+  // Set a seed value for the internal random number generator.
+  void SetSeed(uint64_t seed);
+
+ private:
+  struct SamplingSession {
+    SubsplitDAG& dag_;
+    EigenConstVectorXdRef normalized_sbn_parameters_;
+    EigenConstVectorXdRef inverted_probabilities_;
+    SubsplitDAGStorage result_;
```

Comment:
Interesting that we're using the `SubsplitDAGStorage` here, which I thought was going to be an "implementation detail" of SubsplitDAG. I'm not saying that we shouldn't, but it's worth thinking about how much this is a core data structure in the code.

Reply:
It's not essential as it is only used to build the resulting Node tree, but it reduces the complexity in traversing the sampled nodes. A heavier alternative would be to use a throwaway instance of SubsplitDAG, or using a map with <Node, pair<Node, Node> for the two children of each node. I'm leaving it unchanged for further discussion.